### DISCUSSION TITLE
Get the value from parent parent field

### DISCUSSION THREAD

**Discussion by donbeave** [graphql-java/graphql-java#2370] [Q&A]
Hello!

I have a schema in GraphQL like this:

```graphql
type Query {
  # find a person by name or returns null
  person(name: String!): Person
}

type Person {
  name: String!
  favouriteFoods: [Product!]!
}

type Product {
  sku: String!
  # returns true if a person is allergic to this type of product
  allergic: Boolean!
}
```

And I also have few data fetchers, one for root query: `person(name: String!)` which returns plain Java object without favouriteFoods field, and also one data fetcher for `favouriteFoods: [Product!]!` field in Person type, which returns plain a list of products which this person likes, this list is plain java objects - Product with only one property - sku. So I need to add one data fetched to the field `allergic` in `Product` type, but in my case, this data fetcher needs to have 2 values, one is Person object and another is Product object. The current `graphql-java` DataFetcher implementation is only given one method with a single argument `DataFetchingEnvironment environment`. This DFE object has a method `getSource()` which returns only parent type - `Product`, but I also need somehow to get parent parent type - `Person`. Is it possible to do so with the current implementation? I can see I can get input values from DFE from the `ExecutionStepInfo` hierarchy, but there is no returned value inside. What is the suggested way to get the java object `Person` from that data fetcher?

**Comment by kvfedorov** [comment]
Since 'alergic' is an attribute of particular person, it is totally legit to have person reference in product

**Comment by bbakerman** [comment]
> I can see I can get input values from DFE from the ExecutionStepInfo hierarchy, but there is no returned value inside. 

We chose not to hold values into the ExecutionStepInfo because of the memory aspects that might entail.  IN fact one could argue we keep to much in ExecutionStepInfo in general already.

If you want to pass "context" data down to a child fetcher then we have a mechanism that you can opt into

See https://www.graphql-java.com/blog/deep-dive-data-fetcher-results/ and the section on local context.

You use DataFetcherResult to pass back a context that will be made available to child fields.  You can invent whatever structure you like for this.

**Reply by donbeave** [reply]
It works well in simple relationships like parent and child. But when need pass parent to sub sub sub child will not looks that nice, because As I understand if use DataFetcherResult I will need to pass parent through many children. Another way I think is use local context. But for me the idea to have an API to get the source for the parent context will be more amazing, cleaner from the code perspective. Maybe not put the value inside the ExecutionStepInfo, but provide a method to get the value by ExecutionStepInfo? As I understand the value already processed and stored in the memory, am I correct?

**Reply by bbakerman** [reply]
>  Maybe not put the value inside the ExecutionStepInfo, but provide a method to get the value by ExecutionStepInfo? As I understand the value already processed and stored in the memory, am I correct?

yes and no - the value that is resolved to might still be in memory or it might not.  

if we hold the grand parent values in memory for every step in the tree then we are guaranteed to use the most amount of memory

once a "source" object is no longer used the JVM is free to clean it up.  So in a memory contrained system this might help keep things working

but by creating an API to get grand parent values ALWAYS we move to a maximum memory usage model.

So while the API is less clean (when using local context) those who want it can still do it (as the expense of memory) and those who do not want it always are more memory efficient by default.

**Reply by donbeave** [reply]
Completely agree.
